### ISSUE TITLE
Compiler Bug.

### ISSUE THREAD

**Issue by matthewhammer** [dfinity/motoko#2789]
From here: https://github.com/matthewhammer/motoko-adapton/runs/3626127795#step:4:54

```
Motoko (source 45lx73l4-afs7isrj-wl5wg6k2-p6zskr8q)

Fatal error: exception Lib.Promise.Promise
Raised at Lib.Promise.fulfill in file "lib/lib.ml", line 456, characters 55-68
Called from Lib.Promise.lazy_value in file "lib/lib.ml", line 461, characters 34-50
Called from Codegen__Compile.StackRep.adjust in file "codegen/compile.ml", line 5316, characters 48-75
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7566, characters 14-46
Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
Called from Codegen__InstrList.concat_map in file "codegen/instrList.ml" (inlined), line 135, characters 43-58
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6866, characters 6-50
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6812, characters 10-61
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7492, characters 30-61
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 7495, characters 18-27
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.StackRep.materialize_const_v in file "codegen/compile.ml", line 5283, characters 53-64
Called from Lib.Promise.lazy_value in file "lib/lib.ml", line 461, characters 44-50
Called from Codegen__Compile.StackRep.adjust in file "codegen/compile.ml", line 5316, characters 48-75
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7566, characters 14-46
Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
Called from Codegen__InstrList.concat_map in file "codegen/instrList.ml" (inlined), line 135, characters 43-58
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6866, characters 6-50
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6812, characters 10-61
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7492, characters 30-61
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 7495, characters 18-27
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6813, characters 24-34
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6848, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.StackRep.materialize_const_v in file "codegen/compile.ml", line 5283, characters 53-64
Called from Lib.Promise.lazy_value in file "lib/lib.ml", line 461, characters 44-50
Called from Codegen__Compile.StackRep.adjust in file "codegen/compile.ml", line 5316, characters 48-75
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7566, characters 14-46
Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
Called from Codegen__InstrList.concat_map in file "codegen/instrList.ml" (inlined), line 135, characters 43-58
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6866, characters 6-50
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6812, characters 10-61
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7492, characters 30-61
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 7495, characters 18-27
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.StackRep.materialize_const_v in file "codegen/compile.ml", line 5283, characters 53-64
Called from Lib.Promise.lazy_value in file "lib/lib.ml", line 461, characters 44-50
Called from Codegen__Compile.StackRep.adjust in file "codegen/compile.ml", line 5316, characters 48-75
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7566, characters 14-46
Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
Called from Codegen__InstrList.concat_map in file "codegen/instrList.ml" (inlined), line 135, characters 43-58
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6866, characters 6-50
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6812, characters 10-61
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.def in file "lib/lib.ml", line 499, characters 13-28
Called from Codegen__Compile.compile_dec.(fun) in file "codegen/compile.ml", line 7833, characters 39-50
Called from Codegen__Compile.compile_dec.(fun) in file "codegen/compile.ml", line 7818, characters 30-42
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.(fun) in file "codegen/compile.ml", line 7869, characters 39-61
Called from Codegen__Compile.main_actor.(fun) in file "codegen/compile.ml", line 8064, characters 4-21
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.E.built_in in file "codegen/compile.ml" (inlined), line 401, characters 4-49
Called from Codegen__Compile.IC.export_init.(fun) in file "codegen/compile.ml", line 3453, characters 20-43
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from Codegen__Compile.IC.export_init in file "codegen/compile.ml", line 3450, characters 18-227
Called from Codegen__Compile.compile in file "codegen/compile.ml", line 8167, characters 6-24
Called from Pipeline.compile_files in file "pipeline/pipeline.ml", line 647, characters 14-53
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Moc.process_files in file "exes/moc.ml", line 148, characters 37-82
Called from Moc in file "exes/moc.ml", line 210, characters 4-23

Error: Process completed with exit code 255.
```

**Comment by matthewhammer** [comment]
The line that seems relevant in the code is this one:
https://github.com/matthewhammer/motoko-adapton/commit/ae3321b077b93546222e4c2cf7b880ceb9dd2672

And when I do that edit, the compiler bug is avoided.

**Comment by matthewhammer** [comment]
Update: I had written a theory here about the bug, but it was wrong.  Rather than try to explain, it's simpler just to remove it.

**Comment by matthewhammer** [comment]
Okay, the salient line is the `debug_show` here:
https://github.com/matthewhammer/motoko-adapton/pull/33/commits/defd19481bb049c8e0821e9db7056523082fe824#diff-169edc1e16af97edd4dd926a82b2898d50905ea8e874e158b84f609e3708d661R94

When I remove that expression, the other code avoids the compiler error.

**Comment by nomeata** [comment]
What is this type `log`? Can you reproduce it in isolation (just the type definition and a call to `debug_show`)? Did you already reproduce it with latest master?

**Comment by matthewhammer** [comment]
> What is this type `log`?

Unfortunately, I think the complexity of this type is probably the root of the issue, and also makes it hard to simplify and explain.  But let me try to start.

It's a "generic" log type with four parameters that specialize it for each `example` (so far just two).

The generic type definition is here:
  https://github.com/matthewhammer/motoko-adapton/blob/ed42a2c94d30a127e31029e6784b447b8b48766a/src/types/Log.mo#L16
  
In this example that activates the bug, `sort`, the (four) parameters are given here:
 - Class constructor application, where they are listed (definitions above):
 https://github.com/matthewhammer/motoko-adapton/pull/34/files#diff-169edc1e16af97edd4dd926a82b2898d50905ea8e874e158b84f609e3708d661R37
 - I am guessing (not sure) that the compiler is becoming confused by the way that the definition of `Val` and `Exp` are defined and then used to parameterized by that `Engine class`, here:
   - https://github.com/matthewhammer/motoko-adapton/pull/34/files#diff-169edc1e16af97edd4dd926a82b2898d50905ea8e874e158b84f609e3708d661R20
   - https://github.com/matthewhammer/motoko-adapton/pull/34/files#diff-169edc1e16af97edd4dd926a82b2898d50905ea8e874e158b84f609e3708d661R31
 
> Can you reproduce it in isolation (just the type definition and a call to debug_show)?

No, but if you think it would be necessary, I will try that, certainly.  But see below.

> Did you already reproduce it with latest master?

No --- and I just tried that, but I failed.  Of course, I am very happy to report that!  (Thanks for asking that I try.)

But now I wonder what change (or other fix) fixed the bug.  Locally and in CI, I was using dfx 8.0. 

I just checked and saw that 8.1 exists, and with that, the bug goes away as well.

So something happened between those versions, but it doesn't seem urgent to find out, though I'm slightly curious.  We can close this now, I suppose.

Thanks again.
